File: app/src/main/java/com/pluscubed/logcat/helper/ProcessInfoResolver.java
package com.pluscubed.logcat.helper;

import android.app.ActivityManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.text.TextUtils;

import com.pluscubed.logcat.data.ProcessInfo;
import com.pluscubed.logcat.util.UtilLogger;

import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class ProcessInfoResolver {

    private static final long ACTIVITY_MANAGER_REFRESH_MS = 5_000;
    private static final long ROOT_PS_REFRESH_MS = 30_000;
    private static final ProcessInfoResolver INSTANCE = new ProcessInfoResolver();
    private static final UtilLogger log = new UtilLogger(ProcessInfoResolver.class);

    private final Object lock = new Object();
    private final Map<Integer, ProcessInfo> processMap = new HashMap<>();
    private long lastActivityManagerRefresh;
    private long lastRootPsRefresh;
    private boolean refreshRunning;

    private ProcessInfoResolver() {
    }

    public static ProcessInfoResolver getInstance() {
        return INSTANCE;
    }

    public ProcessInfo resolve(Context context, int pid) {
        if (pid <= 0) {
            return ProcessInfo.UNKNOWN;
        }

        refreshIfNeeded(context.getApplicationContext());

        synchronized (lock) {
            ProcessInfo info = processMap.get(pid);
            return info != null ? info : new ProcessInfo(pid, "", "", "", "", -1);
        }
    }

    private void refreshIfNeeded(final Context context) {
        long now = System.currentTimeMillis();
        synchronized (lock) {
            boolean activityManagerStale = now - lastActivityManagerRefresh >= ACTIVITY_MANAGER_REFRESH_MS;
            boolean rootPsStale = now - lastRootPsRefresh >= ROOT_PS_REFRESH_MS;
            if (refreshRunning || (!activityManagerStale && !rootPsStale)) {
                return;
            }
            refreshRunning = true;
            if (activityManagerStale) {
                lastActivityManagerRefresh = now;
            }
            if (rootPsStale) {
                lastRootPsRefresh = now;
            }
        }

        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    Map<Integer, ProcessInfo> snapshot = new HashMap<>();
                    snapshot.putAll(readActivityManager(context));
                    for (Map.Entry<Integer, ProcessInfo> entry : readRootPs().entrySet()) {
                        snapshot.put(entry.getKey(), merge(snapshot.get(entry.getKey()), entry.getValue()));
                    }
                    synchronized (lock) {
                        for (Map.Entry<Integer, ProcessInfo> entry : snapshot.entrySet()) {
                            ProcessInfo previous = processMap.get(entry.getKey());
                            processMap.put(entry.getKey(), merge(previous, entry.getValue()));
                        }
                    }
                } finally {
                    synchronized (lock) {
                        refreshRunning = false;
                    }
                }
            }
        }, "DaoLogcatProcessInfo").start();
    }

    private Map<Integer, ProcessInfo> readActivityManager(Context context) {
        Map<Integer, ProcessInfo> result = new HashMap<>();
        ActivityManager activityManager =
                (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        if (activityManager == null) {
            return result;
        }

        List<ActivityManager.RunningAppProcessInfo> processes =
                activityManager.getRunningAppProcesses();
        if (processes == null) {
            return result;
        }

        PackageManager packageManager = context.getPackageManager();
        for (ActivityManager.RunningAppProcessInfo process : processes) {
            String packageName = firstPackageName(process.pkgList,
                    packageManager.getPackagesForUid(process.uid));
            result.put(process.pid, new ProcessInfo(process.pid, process.processName, packageName,
                    "", "", process.uid));
        }
        return result;
    }

    private Map<Integer, ProcessInfo> readRootPs() {
        Map<Integer, ProcessInfo> result = new HashMap<>();
        Process process = null;
        PrintStream outputStream = null;
        BufferedReader reader = null;
        try {
            process = Runtime.getRuntime().exec("su");
            outputStream = new PrintStream(new BufferedOutputStream(process.getOutputStream(), 8192));
            outputStream.println("ps -AZ");
            outputStream.println("exit");
            outputStream.flush();
            outputStream.close();
            outputStream = null;

            reader = new BufferedReader(new InputStreamReader(process.getInputStream()), 8192);
            String line;
            while ((line = reader.readLine()) != null) {
                ProcessInfo info = parsePsAzLine(line);
                if (info != null) {
                    result.put(info.getPid(), info);
                }
            }
            process.waitFor();
        } catch (IOException | InterruptedException e) {
            log.w(e, "Cannot refresh process list from root ps");
        } finally {
            if (outputStream != null) {
                outputStream.close();
            }
            if (reader != null) {
                try {
                    reader.close();
                } catch (IOException ignored) {
                }
            }
            if (process != null) {
                process.destroy();
            }
        }
        return result;
    }

    private static ProcessInfo parsePsAzLine(String line) {
        if (TextUtils.isEmpty(line) || line.startsWith("LABEL ")) {
            return null;
        }

        String[] parts = line.trim().split("\\s+");
        if (parts.length < 4) {
            return null;
        }

        int pid;
        try {
            pid = Integer.parseInt(parts[2]);
        } catch (NumberFormatException e) {
            return null;
        }

        StringBuilder name = new StringBuilder(parts[parts.length - 1]);
        return new ProcessInfo(pid, name.toString(), "", parts[1], parts[0], -1);
    }

    private static ProcessInfo merge(ProcessInfo previous, ProcessInfo current) {
        if (previous == null) {
            return current;
        }

        return new ProcessInfo(
                current.getPid() > 0 ? current.getPid() : previous.getPid(),
                nonEmpty(current.getProcessName(), previous.getProcessName()),
                nonEmpty(current.getPackageName(), previous.getPackageName()),
                nonEmpty(current.getUser(), previous.getUser()),
                nonEmpty(current.getSelinuxLabel(), previous.getSelinuxLabel()),
                current.getUid() >= 0 ? current.getUid() : previous.getUid());
    }

    private static String nonEmpty(String preferred, String fallback) {
        return !TextUtils.isEmpty(preferred) ? preferred : fallback;
    }

    private static String firstPackageName(String[] pkgList, String[] uidPackages) {
        if (pkgList != null && pkgList.length > 0) {
            return pkgList[0];
        }
        if (uidPackages != null && uidPackages.length > 0) {
            return uidPackages[0];
        }
        return "";
    }
}

File: app/src/main/java/com/pluscubed/logcat/helper/ProcessInfoResolver.kt
package com.pluscubed.logcat.helper

import android.app.ActivityManager
import android.content.Context
import android.content.pm.PackageManager
import android.text.TextUtils
import com.pluscubed.logcat.data.ProcessInfo
import com.pluscubed.logcat.util.UtilLogger
import java.io.BufferedOutputStream
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintStream
import java.util.HashMap

class ProcessInfoResolver private constructor() {
    private val lock = Any()
    private val processMap: MutableMap<Int, ProcessInfo> = HashMap()
    private var lastActivityManagerRefresh: Long = 0
    private var lastRootPsRefresh: Long = 0
    private var refreshRunning = false

    fun resolve(context: Context, pid: Int): ProcessInfo {
        if (pid <= 0) {
            return ProcessInfo.UNKNOWN
        }
        refreshIfNeeded(context.applicationContext)
        synchronized(lock) {
            val info = processMap[pid]
            return info ?: ProcessInfo(pid, "", "", "", "", -1)
        }
    }

    private fun refreshIfNeeded(context: Context) {
        val now = System.currentTimeMillis()
        synchronized(lock) {
            val activityManagerStale = now - lastActivityManagerRefresh >= ACTIVITY_MANAGER_REFRESH_MS
            val rootPsStale = now - lastRootPsRefresh >= ROOT_PS_REFRESH_MS
            if (refreshRunning || (!activityManagerStale && !rootPsStale)) {
                return
            }
            refreshRunning = true
            if (activityManagerStale) {
                lastActivityManagerRefresh = now
            }
            if (rootPsStale) {
                lastRootPsRefresh = now
            }
        }

        Thread({
            try {
                val snapshot: MutableMap<Int, ProcessInfo> = HashMap()
                snapshot.putAll(readActivityManager(context))
                for ((key, value) in readRootPs()) {
                    snapshot[key] = merge(snapshot[key], value)
                }
                synchronized(lock) {
                    for ((key, value) in snapshot) {
                        val previous = processMap[key]
                        processMap[key] = merge(previous, value)
                    }
                }
            } finally {
                synchronized(lock) {
                    refreshRunning = false
                }
            }
        }, "DaoLogcatProcessInfo").start()
    }

    private fun readActivityManager(context: Context): Map<Int, ProcessInfo> {
        val result: MutableMap<Int, ProcessInfo> = HashMap()
        val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager?
            ?: return result
        val processes = activityManager.runningAppProcesses ?: return result
        val packageManager = context.packageManager
        for (process in processes) {
            val packageName = firstPackageName(
                process.pkgList,
                packageManager.getPackagesForUid(process.uid)
            )
            result[process.pid] = ProcessInfo(
                process.pid, process.processName, packageName,
                "", "", process.uid
            )
        }
        return result
    }

    private fun readRootPs(): Map<Int, ProcessInfo> {
        val result: MutableMap<Int, ProcessInfo> = HashMap()
        var process: Process? = null
        var outputStream: PrintStream? = null
        var reader: BufferedReader? = null
        try {
            process = Runtime.getRuntime().exec("su")
            outputStream = PrintStream(BufferedOutputStream(process.outputStream, 8192))
            outputStream.println("ps -AZ")
            outputStream.println("exit")
            outputStream.flush()
            outputStream.close()
            outputStream = null
            reader = BufferedReader(InputStreamReader(process.inputStream), 8192)
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                val info = parsePsAzLine(line)
                if (info != null) {
                    result[info.pid] = info
                }
            }
            process.waitFor()
        } catch (e: Exception) {
            log.w(e, "Cannot refresh process list from root ps")
        } finally {
            outputStream?.close()
            try {
                reader?.close()
            } catch (ignored: IOException) {
            }
            process?.destroy()
        }
        return result
    }

    private fun parsePsAzLine(line: String?): ProcessInfo? {
        if (TextUtils.isEmpty(line) || line!!.startsWith("LABEL ")) {
            return null
        }
        val parts = line.trim { it <= ' ' }.split("\\s+".toRegex()).toTypedArray()
        if (parts.size < 4) {
            return null
        }
        val pid: Int = try {
            parts[2].toInt()
        } catch (e: NumberFormatException) {
            return null
        }
        val name = parts[parts.size - 1]
        return ProcessInfo(pid, name, "", parts[1], parts[0], -1)
    }

    private fun merge(previous: ProcessInfo?, current: ProcessInfo): ProcessInfo {
        if (previous == null) {
            return current
        }
        return ProcessInfo(
            if (current.pid > 0) current.pid else previous.pid,
            nonEmpty(current.processName, previous.processName),
            nonEmpty(current.packageName, previous.packageName),
            nonEmpty(current.user, previous.user),
            nonEmpty(current.selinuxLabel, previous.selinuxLabel),
            if (current.uid >= 0) current.uid else previous.uid
        )
    }

    private fun nonEmpty(preferred: String?, fallback: String?): String? {
        return if (!TextUtils.isEmpty(preferred)) preferred else fallback
    }

    private fun firstPackageName(pkgList: Array<String>?, uidPackages: Array<String>?): String {
        if (pkgList != null && pkgList.isNotEmpty()) {
            return pkgList[0]
        }
        return if (uidPackages != null && uidPackages.isNotEmpty()) {
            uidPackages[0]
        } else ""
    }

    companion object {
        private const val ACTIVITY_MANAGER_REFRESH_MS = 5000L
        private const val ROOT_PS_REFRESH_MS = 30000L
        private val INSTANCE = ProcessInfoResolver()
        private val log = UtilLogger(ProcessInfoResolver::class.java)

        @JvmStatic
        fun getInstance(): ProcessInfoResolver {
            return INSTANCE
        }
    }
}
